Automattic/wp-calypso
# Post Editor: move the autosave preview URL logic to Redux

The `PostEditStore` has a `_previewUrl` field that maintains the preview URL of the currently edited post and solves a special case related to autosave. This PR converts that logic to Redux.

Published post is regularly autosaved when edited:  the modifications are not saved to the actual post (that's the normal save/update), but to a special slot that doesn't affect the publicly visible post. That slot can be used to restore the autosaved version if a real save is not performed for some reason.

Draft posts are "autosaved", too, but it's not a real autosave -- it's a normal save. That's OK for drafts that are not yet published.

How does that affect preview? Normally, the preview URL uses the `post.URL` field and does some postprocessing on it (map domain, tweak query params...).

If there is a more recent autosaved version, we store and use the `preview_URL` field that was returned as part of response to the autosave REST request. When the post is later saved, we reset the `autosavePreviewUrl` back to `null` and use `post.URL` again.

**Some code changes that deserve an explanation:**
- remove the static call of `wp.undocumented` in `client/state/preferences/actions.js`. The `wp` library is being mocked in the post editor tests and this change removes the need for extra mock.
- the `getPreviewUrl` function in `lib/posts/utils` has a new `autosavePreviewUrl` parameter. That makes more sense than assigning a `preview_URL` property to the post object.

**How to test:**
1. Start editing a published post.
2. Modify the content.
3. Click "Preview" button.
4. Verify that the previewed version contains the modifications.
5. Verify that the published version doesn't contain the modifications yet.
